source/loader/layers/sanitizer/msan/msan_ddi.cpp

```diff
@@ -0,0 +1,1524 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+ * See LICENSE.TXT
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file msan_ddi.cpp
+ *
+ */
+
+#include "msan_ddi.hpp"
+#include "msan_interceptor.hpp"
+#include "sanitizer_common/sanitizer_utils.hpp"
+#include "ur_sanitizer_layer.hpp"
+
+#include <memory>
+
+namespace ur_sanitizer_layer {
+namespace msan {
+
+namespace {
+
+ur_result_t setupContext(ur_context_handle_t Context, uint32_t numDevices,
+                         const ur_device_handle_t *phDevices) {
+    std::shared_ptr<ContextInfo> CI;
+    UR_CALL(getMsanInterceptor()->insertContext(Context, CI));
+    for (uint32_t i = 0; i < numDevices; ++i) {
+        auto hDevice = phDevices[i];
+        std::shared_ptr<DeviceInfo> DI;
+        UR_CALL(getMsanInterceptor()->insertDevice(hDevice, DI));
+        DI->Type = GetDeviceType(Context, hDevice);
+        if (DI->Type == DeviceType::UNKNOWN) {
+            getContext()->logger.error("Unsupport device");
+            return UR_RESULT_ERROR_INVALID_DEVICE;
+        }
+        getContext()->logger.info(
+            "DeviceInfo {} (Type={}, IsSupportSharedSystemUSM={})",
+            (void *)DI->Handle, ToString(DI->Type),
+            DI->IsSupportSharedSystemUSM);
+        getContext()->logger.info("Add {} into context {}", (void *)DI->Handle,
+                                  (void *)Context);
+        if (!DI->Shadow) {
+            UR_CALL(DI->allocShadowMemory(Context));
+        }
+        CI->DeviceList.emplace_back(hDevice);
+        CI->AllocInfosMap[hDevice];
+    }
+    return UR_RESULT_SUCCESS;
+}
+
+bool isInstrumentedKernel(ur_kernel_handle_t hKernel) {
+    auto hProgram = GetProgram(hKernel);
+    auto PI = getMsanInterceptor()->getProgramInfo(hProgram);
+    return PI->isKernelInstrumented(hKernel);
+}
+
+} // namespace
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Intercept function for urAdapterGet
+ur_result_t urAdapterGet(
+    uint32_t
+        NumEntries, ///< [in] the number of adapters to be added to phAdapters.
+    ///< If phAdapters is not NULL, then NumEntries should be greater than
+    ///< zero, otherwise ::UR_RESULT_ERROR_INVALID_SIZE,
+    ///< will be returned.
+    ur_adapter_handle_t *
+        phAdapters, ///< [out][optional][range(0, NumEntries)] array of handle of adapters.
+    ///< If NumEntries is less than the number of adapters available, then
+    ///< ::urAdapterGet shall only retrieve that number of platforms.
+    uint32_t *
+        pNumAdapters ///< [out][optional] returns the total number of adapters available.
+) {
+    auto pfnAdapterGet = getContext()->urDdiTable.Global.pfnAdapterGet;
+
+    setenv("NEOReadDebugKeys", "1", 1);
+    setenv("AllocateHostAllocationsInHeapExtendedHost", "0", 1);
```

Comment:
Could we add some comments here to help other understand why we need these keys?